plugins/file_chooser/gtk/file_chooser_plugin.cc

```diff
@@ -0,0 +1,198 @@
+
+#include <flutter_linux/flutter_linux.h>
+#include <gtk/gtk.h>
+
+// See channel_controller.dart for documentation.
+const char kChannelName[] = "flutter/filechooser";
+const char kShowOpenPanelMethod[] = "FileChooser.Show.Open";
+const char kShowSavePanelMethod[] = "FileChooser.Show.Save";
+const char kInitialDirectoryKey[] = "initialDirectory";
+const char kInitialFileNameKey[] = "initialFileName";
+const char kAllowedFileTypesKey[] = "allowedFileTypes";
+const char kConfirmButtonTextKey[] = "confirmButtonText";
+const char kAllowsMultipleSelectionKey[] = "allowsMultipleSelection";
+const char kCanChooseDirectoriesKey[] = "canChooseDirectories";
+
+struct _FlFileChooserPlugin {
+  GObject parent_instance;
+
+  // Connection to Flutter engine.
+  FlMethodChannel* channel;
+
+  // Window to attach dialog to.
+  GtkWindow* window;
+};
+
+G_DEFINE_TYPE(FlFileChooserPlugin, fl_file_chooser_plugin, G_TYPE_OBJECT)
+
+// Shows the requested dialog type.
+static FlValue* show_dialog(FlFileChooserPlugin* self,
+                            GtkFileChooserAction action, const gchar* title,
+                            const gchar* default_confirm_button_text,
+                            FlValue* properties) {
+  FlValue* value = fl_value_lookup_string(properties, kConfirmButtonTextKey);
+  const gchar* confirm_button_text =
+      fl_value_get_type(value) == FL_VALUE_TYPE_STRING
+          ? fl_value_get_string(value)
+          : default_confirm_button_text;
+
+  g_autoptr(GtkFileChooserNative) dialog =
+      GTK_FILE_CHOOSER_NATIVE(gtk_file_chooser_native_new(
+          title, nullptr, action, confirm_button_text, "_Cancel"));
+  gtk_window_set_transient_for(GTK_WINDOW(dialog), self->window);
+
+  value = fl_value_lookup_string(properties, kAllowsMultipleSelectionKey);
+  if (fl_value_get_type(value) == FL_VALUE_TYPE_BOOL) {
+    gtk_file_chooser_set_select_multiple(GTK_FILE_CHOOSER(dialog),
+                                         fl_value_get_bool(value));
+  }
+
+  value = fl_value_lookup_string(properties, kCanChooseDirectoriesKey);
+  if (fl_value_get_type(value) == FL_VALUE_TYPE_BOOL &&
+      fl_value_get_bool(value)) {
+    gtk_file_chooser_set_action(GTK_FILE_CHOOSER(dialog),
+                                GTK_FILE_CHOOSER_ACTION_SELECT_FOLDER);
+  }
+
+  value = fl_value_lookup_string(properties, kInitialDirectoryKey);
+  if (fl_value_get_type(value) == FL_VALUE_TYPE_STRING) {
+    gtk_file_chooser_set_current_folder(GTK_FILE_CHOOSER(dialog),
+                                        fl_value_get_string(value));
+  }
+
+  value = fl_value_lookup_string(properties, kInitialFileNameKey);
+  if (fl_value_get_type(value) == FL_VALUE_TYPE_STRING) {
+    gtk_file_chooser_set_current_name(GTK_FILE_CHOOSER(dialog),
+                                      fl_value_get_string(value));
+  }
+
+  value = fl_value_lookup_string(properties, kAllowedFileTypesKey);
+  if (fl_value_get_type(value) == FL_VALUE_TYPE_LIST) {
```

Comment:
This block is quite large; let's extract the loop body to a helper that takes `filter_info` and returns `filter`.